src/plugin.ts

```diff
@@ -0,0 +1,199 @@
+   * Shorthand for `typeConfig.extensions.nexus`
+   */
+  typeExtension: NexusTypeExtensions;
+  /**
+   * Nexus specific metadata provided to the field
+   * Shorthand for `fieldConfig.extensions.nexus`
+   */
+  fieldExtension: NexusFieldExtension;
+};
+
+export type StringLike = PrintedTypeGen | string;
+
+export interface PluginConfig {
+  /**
+   * A name for the plugin, useful for errors, etc.
+   */
+  name: string;
+  /**
+   * A description for the plugin
+   */
+  description?: string;
+  /**
+   * Any type definitions we want to add to the field definitions
+   */
+  fieldDefTypes?: StringLike | StringLike[];
+  /**
+   * Any type definitions we want to add to the type definition option
+   */
+  objectTypeDefTypes?: StringLike | StringLike[];
+  /**
+   * Any type definitions we want to add to the schema
+   */
+  schemaDefTypes?: StringLike | StringLike[];
+  /**
+   * Executed once, just before the types are walked. Useful for defining custom extensions
+   * to the "definition" builders that are needed while traversing the type definitions, as
+   * are defined by `dynamicOutput{Method,Property}` / `dynamicInput{Method,Property}`
+   */
+  onInstall?: (builder: BuilderLens) => void;
+  /**
+   * Executed once, just after types have been walked but also before the schema definition
+   * types are materialized into GraphQL types. Use this opportunity to add / modify / remove
+   * any types before we go through the resolution step.
+   */
+  onBeforeBuild?: (builder: BuilderLens) => void;
+  /**
+   * Executed any time a field resolver is created. Returning a function here will add its in the
+   * stack of middlewares with the (root, args, ctx, info, next) signature, where the `next` is the
+   * next middleware or resolver to be executed.
+   */
+  onCreateFieldResolver?: (
+    createResolverInfo: CreateFieldResolverInfo
+  ) => MiddlewareFn | undefined;
+  /**
+   * Executed any time a "subscribe" handler is created. Returning a function here will add its in the
+   * stack of middlewares with the (root, args, ctx, info, next) signature, where the `next` is the
+   * next middleware or resolver to be executed.
+   */
+  onCreateFieldSubscribe?: (
+    createSubscribeInfo: CreateFieldResolverInfo
+  ) => MiddlewareFn | undefined;
+  /**
+   * Executed when a field is going to be printed to the nexus "generated types". Gives
+   * an opportunity to override the standard behavior for printing our inferrred type info
+   */
+  onPrint?: (visitor: Visitor<ASTKindToNode>) => void;
+}
+
+/**
+ * Helper for allowing plugins to fulfill the return of the `next` resolver,
+ * without paying the cost of the Promise if not required.
+ */
+export function completeValue<T>(
+  valOrPromise: PromiseLike<T> | T,
+  onSuccess: (completedVal: T) => T,
+  onError?: (errVal: any) => T
+) {
+  if (isPromiseLike(valOrPromise)) {
+    return valOrPromise.then((completedValue) => {
+      return onSuccess(completedValue);
+    }, onError);
```

Comment:
🚲 🏠
```ts
=> 
  isPromiseLike(valOrPromise)
    ? valOrPromise.then(onSuccess).catch(onError)
    : onSuccess(valOrPromise)
```

Reply:
Nah, we want this to be explicit to the promise api, we don't want `onError` to be the error handler for `onSuccess`, that should be propagated.

Reply:
Ah my form is not equivalent. Coolio